### PR TITLE
Add coherence to sections

### PR DESCRIPTION
**Pull request by l-vincent-l**
**Add coherence in journey' sections**

We want to have coherence between the from (to) of a section and the from (to) of the request.
It also adds coherence between the to (from) of the streetnetwork section and the  from(to) of the first (last) public transport section. Therefore the to(from) of the streetnetwork section is now a stop_point, you still have the adress of the stop point in the place.

We also don't really know from where we start when we start from a area (a stop_area or an admin), thus we added a crow_fly section which starts from the centroid of this area and goes to the first stop_point of the first pt_section.

### CODE REVIEW

**Comment by antoine-de** [comment]
@skywave, It should handle #425 

Is it ok with you ?

File: source/routing/raptor_api.cpp [hunk 1]
```diff
@@ -103,12 +103,33 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
...
-            const auto& departure_stop_point = path.items.front().stop_points.front();
-            // for stop areas, we don't want to display the fallback section if start
-            // from one of the stop area's stop point
-            if (origin.uri != departure_stop_point->stop_area->uri) {
+        if (path.items.empty()) {
+            continue;
+        }
+        /*
+         * For the first section, we can distinguish 2 cases
+         * 1) We start from an area, we will add a crow fly section from the centroid of the area
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp:111]: It should be a bit more precise than 'area', say it is ok for stop_area and admin

File: source/routing/raptor_api.cpp [hunk 2]
```diff
@@ -103,12 +103,33 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
...
+         * */
+
+        if (origin.type == type::Type_e::Admin || origin.type == type::Type_e::StopArea) {
+            if (path.items.front().stop_times.empty()) {
+                continue;
+            }
+            const auto sp_dest = path.items.front().stop_times.front()->journey_pattern_point->stop_point;
+            type::EntryPoint destination(type::Type_e::StopPoint, sp_dest->uri);
+            bt::time_period action_period(path.items.front().departures.front(),
+                                          path.items.front().departures.front()+bt::minutes(1));
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp:126]: why add 1mn ?

**Comment by koch-t** [comment]
Two things, with crow_fly you can't distinguish  walking from bicycle, car, etc. Durations are 0 seconds also?

**Comment by koch-t** [comment]
You now also get crow_fly's if you request from/to the stop_area you are also actually departing from. Something previously turned off explicitly

**Comment by l-vincent-l** [comment]
Crowfly sections are informative, and not precise, you don't really know from where you start, so you don't have a walking or bicycle (actually you're traveling on a crow, how cool is it?!?), so can't really put a duration, maybe we could put an indicative one.

I'm not sure on what to do with your second comment, the thing is it would be easier for web integrator to always have stop_point in public transport section. Also if I put a stop_area in the pt_section and no crow_fly before, you won't have the information about the stop_point you're starting by, and I think you want that information.

**Comment by koch-t** [comment]
There are also situations where you start from Stop A, but it's much faster to start from Station C which is 10 minutes of walking from A. 
With the current code Navitia makes a crow_fly of 0 seconds for that, also there is no longer a instructions/geometry for the walking/biking/car path

**Comment by l-vincent-l** [comment]
Yep, I need to put a duration for this case.

**Comment by l-vincent-l** [comment]
@skywave There's now a duration when we start/go to a stop_area and the first/last stop point of the next/previous pt section is not one of the one of the stop_area.

**Comment by koch-t** [comment]
I still think the mode (walking/bike/car) is missing depending on the first/last section_mode. And i think it would still be a good idea to include geometry/instructions for this section. Except when it is the same stop_area ofcourse.

**Comment by l-vincent-l** [comment]
Jenkins, test this please

File: source/routing/raptor_api.cpp [hunk 3]
```diff
@@ -103,15 +103,47 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
...
+            if (path.items.front().stop_times.empty()) {
+                continue;
+            }
+            const auto sp_dest = path.items.front().stop_times.front()->journey_pattern_point->stop_point;
+            type::EntryPoint destination_tmp(type::Type_e::StopPoint, sp_dest->uri);
+            bt::time_period action_period(path.items.front().departures.front(),
+                                          path.items.front().departures.front()+bt::minutes(1));
+            fill_crowfly_section(origin, destination_tmp, path.items.front().departures.front(),
+                                 d, enhanced_response, pb_journey, now, action_period);
+            // If we start from a stop_area, and the last stop point is not from that stop_area,
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp]: first stop point

File: source/routing/raptor_api.cpp [hunk 4]
```diff
@@ -103,15 +103,47 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
...
+            // we indicate a time.
+            if (origin.type == type::Type_e::StopArea && origin.uri != sp_dest->stop_area->uri) {
+                const auto duration = worker.get_distance(sp_dest->idx);
+                if (!duration.is_pos_infinity() && !duration.is_neg_infinity() && !duration.is_not_a_date_time()) {
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp]: ya une fonction boost pour ca: is_special()

File: source/routing/raptor_api.cpp [hunk 5]
```diff
@@ -120,10 +152,19 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
 
                     const auto walking_time = temp.duration;
                     departure_time = path.items.front().departure - walking_time.to_posix();
-                    fill_street_sections(enhanced_response, origin, temp, d, pb_journey, departure_time);
+                    fill_street_sections(enhanced_response, origin, temp, d, pb_journey,
+                        departure_time);
+                    auto section = pb_journey->mutable_sections(pb_journey->mutable_sections()->size()-1);
+                    bt::time_period action_period(boost::posix_time::from_iso_string(section->begin_date_time()),
+                                                  boost::posix_time::from_iso_string(section->end_date_time()));
+                    // We had coherence with the origin of the request
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp]: we add ?

File: source/routing/raptor_api.cpp [hunk 6]
```diff
@@ -120,10 +152,19 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
...
+                    fill_pb_placemark(origin, d, section->mutable_origin(), 2, now, action_period, show_codes);
+                    // We had coherence with the first pt section
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp]: same

File: source/routing/raptor_api.cpp [hunk 7]
```diff
@@ -120,10 +152,19 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
...
+                    fill_pb_placemark(departure_stop_point, d, section->mutable_destination(), 2, now, action_period, show_codes);
                 }
             }
         }
+        // Here we make origin and departure consistent with the parameters and the first section
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp]: remove comment ?

File: source/routing/raptor_api.cpp [hunk 8]
```diff
@@ -230,24 +271,54 @@ pbnavitia::Response make_pathes(const std::vector<navitia::routing::Path>& paths
...
+            bt::time_period action_period(path.items.back().departures.back(),
+                                          path.items.back().departures.back()+bt::minutes(1));
+            fill_crowfly_section(origin_tmp, destination,path.items.back().departures.back(),
+                                 d, enhanced_response, pb_journey, now, action_period);
+            // If we go to a stop_area, and the last stop point is not from that stop_area,
+            // we indicate a time.
+            if (destination.type == type::Type_e::StopArea &&
+                    destination.uri != sp_orig->stop_area->uri && pb_journey->sections_size() > 0) {
+                auto duration = worker.get_distance(sp_orig->idx, true);
+                if (!duration.is_pos_infinity() && !duration.is_neg_infinity() && !duration.is_not_a_date_time()) {
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp]: same is special

File: source/type/pb_converter.cpp [hunk 9]
```diff
@@ -1257,4 +1304,15 @@ void fill_pb_object(const std::string comment, const nt::Data&,
...
 }
+
+pbnavitia::StreetNetworkMode convert(const navitia::type::Mode_e& mode) {
+    switch (mode) {
+        case navitia::type::Mode_e::Walking : return pbnavitia::Walking;
+        case navitia::type::Mode_e::Bike : return pbnavitia::Bike;
+        case navitia::type::Mode_e::Car : return pbnavitia::Car;
+        case navitia::type::Mode_e::Bss : return pbnavitia::Bss;
+    }
+    return pbnavitia::Walking;
```
Comment (antoine-de) [inline source/type/pb_converter.cpp]: throw no ?

**Pull request merged**